paddle/phi/kernels/gpu/all_gather_gemm_kernel.cu

```diff
@@ -0,0 +1,442 @@
+// Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/phi/common/data_type.h"
+#include "paddle/phi/core/kernel_registry.h"
+
+#ifdef PADDLE_WITH_FLUX
+#include "paddle/phi/backends/dynload/flux.h"
+#endif
+#include "paddle/phi/core/distributed/comm_context_manager.h"
+
+#include "paddle/fluid/distributed/collective/process_group.h"
+#include "paddle/phi/backends/gpu/gpu_context.h"
+#include "paddle/phi/core/dense_tensor.h"
+#include "paddle/phi/kernels/empty_kernel.h"
+#include "paddle/phi/kernels/funcs/math_function.h"
+#include "paddle/phi/kernels/impl/slice_kernel_impl.h"
+#include "paddle/phi/kernels/slice_kernel.h"
+
+#include "paddle/phi/core/distributed/nccl_comm_context.h"
+#include "paddle/phi/core/distributed/utils.h"
+#include "paddle/phi/kernels/gpu/flux_utils.h"
+
+namespace phi {
+
+// Integration of Flux, a gemm-comm-overlap library as described in the paper
+// https://arxiv.org/pdf/2406.06858
+
+#ifdef PADDLE_WITH_FLUX
+template <typename InT, typename OutT>
+class AGGemmHelper {
+ public:
+  static constexpr int MAX_NUM_SIGNAL = 64;
+
+  using FlagType = int32_t;
+  const phi::GPUContext& dev_ctx;
+  distributed::NCCLCommContext* comm_ctx;
+  paddle::distributed::ProcessGroup* tp_group;
+  int32_t nnodes;
+  int32_t full_m;
+  int32_t n_dim;
+  int32_t k_dim;
```

Comment:
Note that when `full_m`, `n_dim` or `k_dim` is larger than `2^31-1`, the operator may crash.